### PR TITLE
DAOS-6808 container: Fix cleanuping resources

### PR DESCRIPTION
**Pull request by dmiter**
- Fix reference counting for container
- Free memory in an error path
- Resolve the race in REBUILD code

### CODE REVIEW

**Review by NiuYawei — APPROVED** [review]

**Comment by dmiter** [comment]
FTEST_pool.PoolCreateTests failed because of DAOS-7016

**Review by jolivier23 — APPROVED** [review]

**Review by Nasf-Fan — APPROVED** [review]

**Review by liuxuezhao — APPROVED** [review]

**Comment by dmiter** [comment]
Hmm. It looks we have an issue with rebuild. My understanding is container reference is decreased while object is used.
`dsc_obj_list_obj() -> dc_obj_shard_open() -> dc_cont_tgt_idx2ptr() -> d_hash_rec_decref() -> Assertion '!zombie || d_list_empty(link)' failed.`

**Review by jolivier23 — CHANGES_REQUESTED** [review]
Since the patch touches container references, we probably need to resolve this before we can land it

**Review by liw — APPROVED** [review]

File: src/object/srv_obj_migrate.c [hunk 1]
```diff
@@ -2539,6 +2539,13 @@ migrate_cont_iter_cb(daos_handle_t ih, d_iov_t *key_iov,
 			break;
 	}
 
+	/* FIXME: keep container handle open until all ULTs completes. */
+	for (rc1 = 0; rc1 < 6; rc1++) {
+		if (tls->mpt_obj_generated_ult == tls->mpt_obj_executed_ult)
+			break;
+		sleep(1);
```
Comment (Nasf-Fan) [inline src/object/srv_obj_migrate.c:2546]: Is 6 second enough? or we need to use "do{}while()"? On the other hand, one second for ULT schedule may be too coarse-grained, dss_sleep() is more general used for such purpose in DAOS.

Comment (NiuYawei) [reply src/object/srv_obj_migrate.c:2546]: sleep(1) won't work here, it'll put the whole xstream(thread) into sleep for 1 second.

If the purpose is to wait for all migrating ULTs exit, ABT_cond could be used here. Should we fix this in a separate patch? @wangdi1 , any thoughts?

**Review by NiuYawei — CHANGES_REQUESTED** [review]

File: src/object/srv_obj_migrate.c (hunk 1 above)
Comment (dmiter) [reply src/object/srv_obj_migrate.c]: Re-implemented this more correctly. Now it check the status of rebuild progress from all xstreams.

**Review by Nasf-Fan — APPROVED** [review]

**Review by NiuYawei — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

**Comment by dmiter** [comment]
It looks we still have an issue with rebuild. My understanding is container handle is accessed after it was released.

File: src/object/srv_obj_migrate.c (hunk 1 above)
Comment (wangdi1) [reply src/object/srv_obj_migrate.c]: Hmm, so close container before migrating object might cause issue?  Agree with niu and Nasf, this is too hacky

Also migrate_cont_iter_cb() happens in xstream 0, and obj_generated_ult is increased in xstream 0, but obj_exected_ult will be increased in other ULTs, see migrate_obj_ult().  
.

File: src/object/srv_obj_migrate.c [hunk 2]
```diff
@@ -2539,6 +2591,34 @@ migrate_cont_iter_cb(daos_handle_t ih, d_iov_t *key_iov,
...
+		mq_arg.obj_generated_ult += tls->mpt_obj_generated_ult;
+		if (mq_arg.obj_generated_ult == 0)
+			continue;
+
+		if (mq_arg.obj_generated_ult == mq_arg.obj_executed_ult &&
+		    mq_arg.generated_ult == mq_arg.executed_ult)
+			break;
+	}
+
+	ABT_mutex_free(&mq_arg.status_lock);
```
Comment (wangdi1) [inline src/object/srv_obj_migrate.c:2620]: Sorry, IMHO, this is a bit hacky. I think we should add reference count into dsc_cont_open() and dsc_cont_close(), and does not wait and do collective operation. 

@dmiter  Could we fix  this in a separate patch?

Comment (dmiter) [reply src/object/srv_obj_migrate.c]: Elaborated this solution one more time.

**Review by jolivier23 — DISMISSED** [review]

File: src/object/srv_obj_migrate.c [hunk 3]
```diff
@@ -2517,14 +2522,19 @@ migrate_cont_iter_cb(daos_handle_t ih, d_iov_t *key_iov,
 	 * Open the remote container as a *client* to be used later to pull
 	 * objects
 	 */
+	if (daos_handle_is_valid(tls->mpt_cont_hdl))
+		dsc_cont_close(tls->mpt_pool_hdl, tls->mpt_cont_hdl);
```
Comment (wangdi1) [inline src/object/srv_obj_migrate.c:2526]: Hmm, this can not guarantee all object migration are finished?   Probably you can just do dsc_cont_open/close in migrate_obj_ult(), since cont_iter_cb() is on xstream 0, and migrate_obj_ult() might be in different ult. 

We can see if we can optimize this later.

Comment (dmiter) [reply src/object/srv_obj_migrate.c:2526]: TLS destroyed when all ULTs complete. So, this should guarantee all object migration are finished.

[At the end](https://github.com/daos-stack/daos/blob/a665f2b27e423921d1f44fca63538d1f832013bd/src/object/srv_obj_migrate.c#L2324) of migrate_obj_ult() you can see migrate_pool_tls_put(tls) which will close pool and container.

Comment (wangdi1) [reply src/object/srv_obj_migrate.c:2526]: I mean when the dsc_cont_close(), the dc_cont will be freed.  but how do you know all those object migration ult (migrate_obj_ult) are done before dsc_cont_close.

IMHO, you probably need increase refcount for cont_hdl(dc_cont) before create migrate_obj_ult(), then release the refcount when migrate_obj_ult() is done.

**Comment by ashleypittman** [comment]
Hijacking this PR to say this one dc_cont_alloc() leak is very common, looking at the latest NLT runs from master, the test runs for a couple of minutes and only sends fewer than 400 rpcs in total, yet this leak occurs 692 times.

For those familiar with this code is that expected?  It seems very high indeed.

Data from here.

https://build.hpdd.intel.com/blue/rest/organizations/jenkins/pipelines/daos-stack/pipelines/daos/branches/master/runs/4171/nodes/971/steps/1133/log/?start=0

`[2021-03-25T03:58:26.638Z] Running log_test on /tmp/dnt_server_bbbboptn.log 161.5MiB
[2021-03-25T03:58:41.524Z] Opcode State Transition Tally
[2021-03-25T03:58:41.524Z]     OPCODE    ALLOCATED    SUBMITTED    SENT    COMPLETED    DEALLOCATED    -DER_SUCCESS
[2021-03-25T03:58:41.524Z] ----------  -----------  -----------  ------  -----------  -------------  --------------
[2021-03-25T03:58:41.524Z]  0x1020003           28            0       0            0             28
[2021-03-25T03:58:41.524Z]  0x1020006            1            0       0            1              1               1
[2021-03-25T03:58:41.524Z]  0x102000a            1            0       0            1              1               1
[2021-03-25T03:58:41.524Z]  0x2020000            2            1       1            1              2               1
[2021-03-25T03:58:41.524Z]  0x2020001           28            0       0            0             28
[2021-03-25T03:58:41.524Z]  0x2020002           28            0       0            0             28
[2021-03-25T03:58:41.524Z]  0x2020003           10            0       0            0             10
[2021-03-25T03:58:41.524Z]  0x2020014           28            0       0           28             28              28
[2021-03-25T03:58:41.524Z]  0x2020015            6            0       0            6              6               6
[2021-03-25T03:58:41.524Z]  0x3020000           11            0       0            0             11
[2021-03-25T03:58:41.524Z]  0x3020001            1            0       0            0              1
[2021-03-25T03:58:41.524Z]  0x3020002           43            0       0            0             43
[2021-03-25T03:58:41.524Z]  0x3020003           39            0       0            0             39
[2021-03-25T03:58:41.524Z]  0x3020004           34            0       0            0             34
[2021-03-25T03:58:41.524Z]  0x3020005           30            0       0            0             30
[2021-03-25T03:58:41.524Z]  0x3020011            1            0       0            1              1               1
[2021-03-25T03:58:41.524Z]  0x3020012            2            0       0            2              2               2
[2021-03-25T03:58:41.524Z]  0x4030000          278            0       0            0            278
[2021-03-25T03:58:41.524Z]  0x4030001         1576            0       0            0           1576
[2021-03-25T03:58:41.524Z]  0x4030002          178            0       0            0            178
[2021-03-25T03:58:41.524Z]  0x4030003            5            0       0            0              5
[2021-03-25T03:58:41.524Z]  0x4030007           13            0       0            0             13
[2021-03-25T03:58:41.524Z]  0x4030008           13            0       0            0             13
[2021-03-25T03:58:41.524Z]  0x4030009         2276            0       0            0           2276
[2021-03-25T03:58:41.524Z]  0x4030012           28           12      12           12             28              12
[2021-03-25T03:58:41.524Z]  0x6020000            1            0       0            1              1               1
[2021-03-25T03:58:41.524Z]  0xa010000           24           12      12           12             24              12
[2021-03-25T03:58:41.524Z] 0xf1020001            1            0       0            0              1
[2021-03-25T03:58:41.524Z] 0xf4010002          307            0       0          307            307             307
[2021-03-25T03:58:41.524Z] 0xff040001            1            0       0            0              1
[2021-03-25T03:58:41.524Z] 0xff040002            1            0       0            0              1
[2021-03-25T03:58:41.524Z] src/vos/vos_space.c:58:1: NORMAL: warning in strict mode 'vos_space_sys_init() Disable SCM space reserving for tiny pool:00000000 sys[2190450688] > tot[134217728]'
[2021-03-25T03:58:42.942Z] src/cart/crt_group.c:3189:1: NORMAL: warning in strict mode 'crt_group_mod_get() Membership unchanged. No modification performed'
[2021-03-25T03:58:42.942Z] src/vos/vos_space.c:58:1: NORMAL: warning in strict mode 'vos_space_sys_init() Disable SCM space reserving for tiny pool:f9f1a2e7 sys[2190450688] > tot[2335750000]'
[2021-03-25T03:58:42.942Z] src/rdb/rdb_raft.c:1664:1: NORMAL: warning in strict mode 'rdb_raft_step_up() uuid/rank: became leader of term 0'
[2021-03-25T03:58:42.942Z] src/pool/srv_target.c:1241:1: NORMAL: warning in strict mode 'ds_pool_tgt_map_update() Ignore update pool uuid 1 -> 1'
[2021-03-25T03:59:09.542Z] src/rdb/rdb_raft.c:1687:1: NORMAL: warning in strict mode 'rdb_raft_step_down() uuid/rank: no longer leader of term 0'
[2021-03-25T03:59:09.542Z] src/rdb/rdb.c:415:1: NORMAL: warning in strict mode 'rdb_stop() db is NULL'
[2021-03-25T03:59:09.542Z] Pid 33386, 1223499 lines total, 28089 trace (2.30%)
[2021-03-25T03:59:09.542Z] Memsize: Total:155,008 HWM:1,925,459,427 221274 allocations, 220582 frees 692 possible leaks
[2021-03-25T03:59:09.542Z] src/container/cli.c:435:1: NORMAL: memory not freed 'dc_cont_alloc() alloc(calloc) 'dc': 224 at 0x....'
[2021-03-25T03:59:31.690Z] Parsed 1223499 lines of logs
[2021-03-25T03:59:31.690Z] Most common logging locations
[2021-03-25T03:59:31.690Z] Logging used 82344 times at src/common/btree.c:1298 (6.7%)
[2021-03-25T03:59:31.690Z] Logging used 65536 times at src/engine/sched.c:397 (5.4%)
[2021-03-25T03:59:31.690Z] Logging used 65536 times at src/engine/sched.c:386 (5.4%)
[2021-03-25T03:59:31.690Z] Logging used 47989 times at src/common/btree.c:347 (3.9%)
[2021-03-25T03:59:31.690Z] Logging used 41278 times at src/common/btree.c:1368 (3.4%)
[2021-03-25T03:59:31.690Z] Logging used 39773 times at src/container/srv_target.c:205 (3.3%)
[2021-03-25T03:59:31.690Z] Logging used 34269 times at src/vos/vos_ilog.c:231 (2.8%)
[2021-03-25T03:59:31.690Z] Logging used 31284 times at src/common/btree.c:1582 (2.6%)
[2021-03-25T03:59:31.690Z] Logging used 20407 times at src/common/mem.c:543 (1.7%)
[2021-03-25T03:59:31.690Z] Logging used 18371 times at src/common/btree.c:279 (1.5%)
`

File: src/object/srv_obj_migrate.c [hunk 4]
```diff
@@ -2366,7 +2372,8 @@ migrate_obj_ult(void *data)
 	rc = migrate_one_epoch_object(oh, &epr, tls, arg);
 
 close:
-	dsc_obj_close(oh);
+	/* FIXME: should keep this handle open until all ULTs complete:
+	dsc_obj_close(oh);*/
```
[the remark this answers is not in the document]
Comment (dmiter) [reply src/object/srv_obj_migrate.c:2376]: This leak is to check the issue here only. My assumption is we close this handle before all ULTs complete.

File: src/object/srv_obj_migrate.c (hunk 3 above)
Comment (dmiter) [reply src/object/srv_obj_migrate.c:2585]: Probably you right. Maybe we should keep it open until completing all ULTs. The same looks for object handle. Need to elaborate this one more time.

Comment (NiuYawei) [reply src/object/srv_obj_migrate.c:2585]: Are we going to fix it in current patch or a follow-on patch?

Comment (dmiter) [reply src/object/srv_obj_migrate.c]: @NiuYawei, I have to resolve somehow rebuild test fails. Otherwise this patch will not be landed.

File: src/object/srv_obj_migrate.c [hunk 5]
```diff
@@ -1308,33 +1308,28 @@ migrate_dkey(struct migrate_pool_tls *tls, struct migrate_one *mrone,
...
 	if (rc)
-		D_GOTO(free, rc);
+		D_GOTO(out, rc);
 
-	/* Open client dc handle used to read the remote object data */
-	rc = dsc_cont_open(tls->mpt_pool_hdl, mrone->mo_cont_uuid,
-			   tls->mpt_coh_uuid, 0, &coh);
+	rc = dsc_pool_open(tls->mpt_pool_uuid, tls->mpt_poh_uuid, 0,
+			   NULL, tls->mpt_pool->spc_pool->sp_map,
+			   &tls->mpt_svc_list, &poh);
```
[the remark this answers is not in the document]
Comment (dmiter) [reply src/object/srv_obj_migrate.c:1323]: I don't store pool handle into TLS. I open it each time and then close. I agree this is not optimal but at least for now I'd like to see there is no race condition any more with this approach.

File: src/object/srv_obj_migrate.c [hunk 6]
```diff
@@ -2042,13 +2038,36 @@ migrate_one_epoch_object(daos_handle_t oh, daos_epoch_range_t *epr,
...
 		DF_U64"-"DF_U64"\n", DP_UOID(arg->oid), arg->shard, epr->epr_lo,
 		epr->epr_hi);
 
+	rc = dsc_pool_open(tls->mpt_pool_uuid, tls->mpt_poh_uuid, 0,
+			   NULL, tls->mpt_pool->spc_pool->sp_map,
+			   &tls->mpt_svc_list, &poh);
+	if (rc) {
+		D_ERROR("dsc_pool_open failed: "DF_RC"\n", DP_RC(rc));
+		D_GOTO(out, rc);
+	}
```
[the remark this answers is not in the document]
Comment (dmiter) [reply src/object/srv_obj_migrate.c:2057]: I don't see that rebuild finish after pool close. It should be reference counted and I just close one instance.

**Comment by dmiter** [comment]
It looks I was able to resolve rebuild race. The solution is not optimal but workable. Please, review it.

**Review by jolivier23 — APPROVED** [review]

**Review by wangdi1 — APPROVED** [review]

File: src/container/container_iv.c [hunk 7]
```diff
@@ -355,8 +355,7 @@ cont_iv_prop_ent_create(struct ds_iv_entry *entry, struct ds_iv_key *key)
 	if (iv_entry == NULL)
 		D_GOTO(out, rc = -DER_NOMEM);
 
-	memcpy(&root_hdl, entry->iv_value.sg_iovs[0].iov_buf,
-		sizeof(root_hdl));
+	memcpy(&root_hdl, entry->iv_value.sg_iovs[0].iov_buf, sizeof(root_hdl));
```
Comment (ashleypittman) [inline src/container/container_iv.c:358]: Whitespace only change

File: src/container/srv_cli.c [hunk 8]
```diff
@@ -103,8 +104,7 @@ dsc_cont_open(daos_handle_t poh, uuid_t cont_uuid, uuid_t coh_uuid,
 		D_GOTO(out, rc = -DER_NOMEM);
 
 	/** destroyed in dsc_cont_close */
-	rc = dsc_cont_csummer_init(&cont->dc_csummer, pool->dp_pool,
-				   cont_uuid);
+	rc = dsc_cont_csummer_init(&cont->dc_csummer, pool->dp_pool, cont_uuid);
```
Comment (ashleypittman) [inline src/container/srv_cli.c:107]: Again, this is whitespace only isn't it?

File: src/gurt/hash.c [hunk 9]
```diff
@@ -1265,6 +1265,10 @@ d_hhash_link_lookup(struct d_hhash *hhash, uint64_t key)
 	if (d_hhash_key_isptr(key)) {
 		struct d_hlink *hlink = (struct d_hlink *)key;
 
+		if (hlink == NULL) {
+			D_ERROR("NULL PTR type key.\n");
+			return NULL;
+		}
```
Comment (ashleypittman) [inline src/gurt/hash.c:1271]: Is this change intentional or left-over debugging?

File: src/object/srv_obj_migrate.c [hunk 10]
```diff
@@ -291,10 +289,8 @@ migrate_pool_tls_destroy(struct migrate_pool_tls *tls)
 		DP_UUID(tls->mpt_pool_uuid), tls->mpt_version);
 	if (tls->mpt_pool)
 		ds_pool_child_put(tls->mpt_pool);
-
```
Comment (ashleypittman) [inline src/object/srv_obj_migrate.c:294]: More whitespace only changes.

**Review by ashleypittman — APPROVED** [review]
Good work, none of my comments are critical.

We'll want this on the 1.2 branch ideally presumably?

File: src/container/srv_cli.c (hunk 8 above)
Comment (dmiter) [reply src/container/srv_cli.c:107]: yes, whitespace only change.

File: src/container/container_iv.c (hunk 7 above)
Comment (dmiter) [reply src/container/container_iv.c:358]: yes, whitespace only change.

File: src/gurt/hash.c (hunk 9 above)
Comment (dmiter) [reply src/gurt/hash.c:1271]: This is for robust error handling. Instead of crashing it should print an error message.

File: src/object/srv_obj_migrate.c (hunk 10 above)
Comment (dmiter) [reply src/object/srv_obj_migrate.c:294]: yes

**Comment by dmiter** [comment]
> We'll want this on the 1.2 branch ideally presumably?

I'd like to land it in 1.2 also. It resolve container leak and race in rebuild code.
